File: packages/sdk-ts/src/core/base32.ts
/**
 * RFC 4648 base32, lowercase, no padding.
 *
 * Used for every stable identifier in the system — content IDs, identity IDs, channel
 * IDs, server IDs (Plans/02 §7). Lowercase and unpadded so an ID survives being read
 * aloud, written on a poster, or typed into a manual node-entry field during a blackout
 * (TP-19), where a `=` character or mixed case is a transcription error waiting to happen.
 */

const ALPHABET = 'abcdefghijklmnopqrstuvwxyz234567';

const DECODE = (() => {
  const table = new Int8Array(128).fill(-1);
  for (let i = 0; i < ALPHABET.length; i++) {
    const c = ALPHABET.charCodeAt(i);
    table[c] = i;
    // Accept uppercase on input — humans transcribing an ID should not be punished
    // for shift key state. Output is always lowercase.
    table[c - 32] = i;
  }
  return table;
})();

export function base32Encode(data: Uint8Array): string {
  let out = '';
  let bits = 0;
  let value = 0;

  for (const byte of data) {
    value = (value << 8) | byte;
    bits += 8;
    while (bits >= 5) {
      out += ALPHABET[(value >>> (bits - 5)) & 31];
      bits -= 5;
    }
  }
  if (bits > 0) out += ALPHABET[(value << (5 - bits)) & 31];

  return out;
}

export function base32Decode(s: string): Uint8Array {
  const out = new Uint8Array(Math.floor((s.length * 5) / 8));
  let bits = 0;
  let value = 0;
  let written = 0;

  for (let i = 0; i < s.length; i++) {
    const code = s.charCodeAt(i);
    const digit = code < 128 ? DECODE[code] : -1;
    if (digit === undefined || digit < 0) {
      throw new Error(`base32: invalid character ${JSON.stringify(s[i])} at index ${i}`);
    }
    value = (value << 5) | digit;
    bits += 5;
    if (bits >= 8) {
      out[written++] = (value >>> (bits - 8)) & 0xff;
      bits -= 8;
    }
  }

  return out.subarray(0, written);
}

File: crates/jb-core/src/base32.rs
//! RFC 4648 base32, lowercase, no padding.
//!
//! Lowercase and unpadded so an identifier survives being read aloud, written on a
//! poster, or typed into a manual node-entry field during a blackout (TP-19).

const ALPHABET: &[u8; 32] = b"abcdefghijklmnopqrstuvwxyz234567";

pub fn encode(data: &[u8]) -> String {
    let mut out = String::with_capacity(data.len().div_ceil(5) * 8);
    let mut bits: u32 = 0;
    let mut value: u32 = 0;

    for &byte in data {
        value = (value << 8) | u32::from(byte);
        bits += 8;
        while bits >= 5 {
            out.push(ALPHABET[((value >> (bits - 5)) & 31) as usize] as char);
            bits -= 5;
        }
    }
    if bits > 0 {
        out.push(ALPHABET[((value << (5 - bits)) & 31) as usize] as char);
    }

    out
}

pub fn decode(s: &str) -> Option<Vec<u8>> {
    let mut out = Vec::with_capacity(s.len() * 5 / 8);
    let mut bits: u32 = 0;
    let mut value: u32 = 0;

    for c in s.chars() {
        // Accept uppercase on input — a person transcribing an ID should not be punished
        // for shift key state. Output is always lowercase.
        let digit = match c.to_ascii_lowercase() {
            'a'..='z' => c.to_ascii_lowercase() as u32 - 'a' as u32,
            '2'..='7' => c as u32 - '2' as u32 + 26,
            _ => return None,
        };
        value = (value << 5) | digit;
        bits += 5;
        if bits >= 8 {
            out.push(((value >> (bits - 8)) & 0xff) as u8);
            bits -= 8;
        }
    }

    Some(out)
}
